util/topsql/reporter/reporter.go

```diff
@@ -209,21 +209,94 @@ func (tsr *RemoteTopSQLReporter) Close() {
 	tsr.client.Close()
 }
 
+// collectedData stores the collected top-sql data.
+type collectedData struct {
+	// data stores topN records for each round.
+	data map[string]*dataPoints
+	// others stores the summary of all evicted records.
+	others *dataPoints
+}
+
+func newCollectedData() *collectedData {
+	return &collectedData{
+		data:   make(map[string]*dataPoints),
+		others: &dataPoints{},
+	}
+}
+
+func (c *collectedData) addEvictedCPUTime(timestamp uint64, totalCPUTimeMs uint32) {
+	if len(c.others.TimestampList) == 0 {
+		c.others.TimestampList = []uint64{timestamp}
+		c.others.CPUTimeMsList = []uint32{totalCPUTimeMs}
+	} else {
+		c.others.TimestampList = append(c.others.TimestampList, timestamp)
+		c.others.CPUTimeMsList = append(c.others.CPUTimeMsList, totalCPUTimeMs)
+	}
+	c.others.CPUTimeMsTotal += uint64(totalCPUTimeMs)
+}
+
+func (d *dataPoints) merge(other *dataPoints) {
+	if other == nil || len(other.TimestampList) == 0 {
+		return
+	}
+	if len(d.TimestampList) == 0 {
+		d.TimestampList = other.TimestampList
+		d.CPUTimeMsList = other.CPUTimeMsList
+		d.CPUTimeMsTotal = other.CPUTimeMsTotal
+		return
+	}
+	d.CPUTimeMsTotal += other.CPUTimeMsTotal
+	findIdx := 0
+	for idx, ts := range other.TimestampList {
+		cpuTimeMs := other.CPUTimeMsList[idx]
+		for ; findIdx < len(d.TimestampList); findIdx++ {
```

Comment:
This relies on Timestamp to never go back, which is not true.

Reply:
nice catch.